core/src/util/shape.rs

```diff
@@ -0,0 +1,243 @@
+use std::fmt;
+
+use crate::ast;
+
+/// Get the "shape" of a fields container, such as a struct or variant.
+pub trait AsShape {
+    fn as_shape(&self) -> Shape;
+}
+
+impl<T> AsShape for ast::Fields<T> {
+    fn as_shape(&self) -> Shape {
+        match self.style {
+            ast::Style::Tuple if self.fields.len() == 1 => Shape::Newtype,
+            ast::Style::Tuple => Shape::Tuple,
+            ast::Style::Struct => Shape::Named,
+            ast::Style::Unit => Shape::Unit,
+        }
+    }
+}
+
+impl AsShape for syn::Fields {
+    fn as_shape(&self) -> Shape {
+        match self {
+            syn::Fields::Named(fields) => fields.as_shape(),
+            syn::Fields::Unnamed(fields) => fields.as_shape(),
+            syn::Fields::Unit => Shape::Unit,
+        }
+    }
+}
+
+impl AsShape for syn::FieldsNamed {
+    fn as_shape(&self) -> Shape {
+        Shape::Named
+    }
+}
+
+impl AsShape for syn::FieldsUnnamed {
+    fn as_shape(&self) -> Shape {
+        if self.unnamed.len() == 1 {
+            Shape::Newtype
+        } else {
+            Shape::Tuple
+        }
+    }
+}
+
+impl AsShape for syn::DataStruct {
+    fn as_shape(&self) -> Shape {
+        self.fields.as_shape()
+    }
+}
+
+impl AsShape for syn::Variant {
+    fn as_shape(&self) -> Shape {
+        self.fields.as_shape()
+    }
+}
+
+/// Description of how fields in a struct or variant are syntactically laid out.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Shape {
+    /// A set of named fields, e.g. `{ field: String }`.
+    Named,
+    /// A list of unnamed fields, e.g. `(String, u64)`.
+    Tuple,
+    /// No fields, e.g. `struct Example;`
+    Unit,
+    /// A special case of [`Tuple`](Shape#variant.Tuple) with exactly one field, e.g. `(String)`.
+    Newtype,
+}
+
+impl Shape {
+    pub fn description(&self) -> &'static str {
+        match self {
+            Shape::Named => "named fields",
+            Shape::Tuple => "unnamed fields",
+            Shape::Unit => "no fields",
+            Shape::Newtype => "one unnamed field",
+        }
+    }
+}
+
+impl AsShape for Shape {
+    fn as_shape(&self) -> Shape {
+        *self
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ShapeSet {
```

Comment:
Should this provide its own `impl FromMeta` for convenience if people want to use it in their crates?

Reply:
Decided that doing this is not a must-have for the initial release, and adding trait impls now is a breaking change to undo later.